plutus-core/plutus-ir/src/PlutusIR/Transform/RewriteRules/Common.hs

```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTs             #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PatternSynonyms   #-}
+{-# LANGUAGE TypeOperators     #-}
+{-# OPTIONS_GHC -Wno-missing-signatures #-}
+
+module PlutusIR.Transform.RewriteRules.Common
+    ( seqA
+    , seqPure
+    , seq
+    , strictLet
+    , pattern A
+    , pattern B
+    , pattern I
+    ) where
+
+import Prelude hiding (seq)
+
+import PlutusCore.Builtin
+import PlutusCore.Quote
+import PlutusIR
+import PlutusIR.Analysis.Builtins
+import PlutusIR.Analysis.VarInfo
+import PlutusIR.MkPir
+import PlutusIR.Purity
+
+{- | Another wrapper that can be more easily turned into an infix operator
+
+e.g. `infixr 0 (***) = seqA binfo vInfo`
+-}
+seqA :: (MonadQuote m, Monoid a, ToBuiltinMeaning uni fun)
+     => BuiltinsInfo uni fun
+     -> VarsInfo tyname Name uni a
+     -> (Type tyname uni a, Term tyname Name uni fun a)
+     ->  m (Term tyname Name uni fun a)
+     -> m (Term tyname Name uni fun a)
+seqA binfo vinfo (a,aT) y = seqPure binfo vinfo a aT <*> y
+
+-- | A thin wrapper to optimize codegen by omitting call to `seq` if left operand `isPure`.
+seqPure :: ( MonadQuote m
+       , Monoid a
+       , ToBuiltinMeaning uni fun
+       , t ~ Term tyname Name uni fun a
+       )
+     => BuiltinsInfo uni fun
+     -> VarsInfo tyname Name uni a
+     -> Type tyname uni a -- ^ the type of left operand a
+     -> t -- ^ left operand a
+     -> m (t -> t) -- ^ how to modify right operand b
+seqPure binfo vinfo aT a | isPure binfo vinfo a = pure id
+                         | otherwise = seq aT a
```

Comment:
You only use `seq` here, right? I'd perhaps inline it then just to reduce the number of options to consider. If somebody needs `seq` but not `seqPure`, it'll be easy for them to bring it back.